### ISSUE TITLE
Invalid OpenTelemetry Trace data can cause Data Prepper to stop ingestion

### ISSUE THREAD

**Issue by dlvenable** [opensearch-project/data-prepper#2185]
@KarstenSchnitter and I encountered another case for otel data crashing data-prepper threads (same impact as above). This time it happens for otel traces. When a span comes with a `links` array and one link misses the `traceState` field, data-prepper produces an error log and the current worker crashes. I am not quite sure if this is invalid input data as above or if this is even valid input data.

**To Reproduce**
We encountered the issue using the [opentelemetry-demo](https://github.com/open-telemetry/opentelemetry-demo) with the _loadgenerator_. Same as above, we provide an alternative approach here, that generates the problematic span directly.

1. Run DataPrepper with the tracing pipeline setup (otel-trace-pipeline, raw-pipeline, service-map-pipeline)
2. Store the following OTEL trace span as JSON in a file trace-request.json. Note, that there is no `traceState` field in the `links` section.
```
{
    "resourceSpans": [
        {
            "resource": {
                "attributes": [
                    {
                        "key": "service.name",
                        "value": {
                            "stringValue": "frontend"
                        }
                    }
                ]
            },
            "scopeSpans": [
                {
                    "scope": {
                        "name": "frontend"
                    },
                    "spans": [
                        {
                            "traceId": "428264014a59a9a29b7053279f687e9f",
                            "spanId": "9bc01dfad9f631ff",
                            "parentSpanId": "",
                            "name": "HTTP GET",
                            "kind": 2,
                            "startTimeUnixNano": "1673448552000000000",
                            "endTimeUnixNano": "1673448552000000100",
                            "attributes": [],
                            "links": [
                                {
                                    "traceId": "c0b1baf0afb0c766a31c5f96619d660e",
                                    "spanId": "d96ff4a5bfe1be72"
                                }
                            ],
                            "status": {}
                        }
                    ]
                }
            ]
        }
    ]
}
```
3. Use the otel-collector just as described in the issue description. Include the otlp trace exporting in its configuration:
```
receivers:
  otlp:
    protocols:
      grpc:
      http:
        cors:
          allowed_origins:
            - "*"

processors:
  batch:
    send_batch_size: 50
    timeout: 1s

exporters:
  logging:
    logLevel: debug
  otlp/traces:
    endpoint: "<DataPrepper Endpoint>"
    tls:
      insecure_skip_verify: true
service:
  pipelines:
    traces:
      receivers: [otlp]
      processors: [batch]
      exporters: [logging, otlp/traces]
```
4. Run the otel-collector like described in the issue description.
5. Send the span to via cURL to the otel-collector which exports it to data-prepper:
`curl -X POST -H "Content-Type: application/json" -d @trace-request.json http://localhost:4318/v1/traces
`
Data-prepper should show the following error:
```
2023-01-13T09:31:33,837 [pool-8-thread-1] ERROR org.opensearch.dataprepper.plugins.source.oteltrace.OTelTraceGrpcService - Failed to parse the request content [resource_spans {
  resource {
    attributes {
      key: "service.name"
      value {
        string_value: "frontend"
      }
    }
  }
  instrumentation_library_spans {
    instrumentation_library {
      name: "frontend"
    }
    spans {
      trace_id: "B\202d\001JY\251\242\233pS\'\237h~\237"
      span_id: "\233\300\035\372\331\3661\377"
      name: "HTTP GET"
      kind: SPAN_KIND_SERVER
      start_time_unix_nano: 1673448552000000000
      end_time_unix_nano: 1673448552000000100
      links {
        trace_id: "\300\261\272\360\257\260\307f\243\034_\226a\235f\016"
        span_id: "\331o\364\245\277\341\276r"
      }
      status {
      }
    }
  }
}
] due to:
java.lang.IllegalArgumentException: traceState cannot be an empty String
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:145) ~[guava-31.1-jre.jar:?]
	at org.opensearch.dataprepper.model.trace.DefaultLink.<init>(DefaultLink.java:41) ~[data-prepper-api-2.0.1.jar:?]
	at org.opensearch.dataprepper.model.trace.DefaultLink$Builder.build(DefaultLink.java:155) ~[data-prepper-api-2.0.1.jar:?]
	at org.opensearch.dataprepper.plugins.otel.codec.OTelProtoCodec$OTelProtoDecoder.getLink(OTelProtoCodec.java:188) ~[otel-proto-common-2.0.1.jar:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197) ~[?:?]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682) ~[?:?]
	at org.opensearch.dataprepper.plugins.otel.codec.OTelProtoCodec$OTelProtoDecoder.parseSpan(OTelProtoCodec.java:122) ~[otel-proto-common-2.0.1.jar:?]
	at org.opensearch.dataprepper.plugins.otel.codec.OTelProtoCodec$OTelProtoDecoder.parseResourceSpans(OTelProtoCodec.java:92) ~[otel-proto-common-2.0.1.jar:?]
	at org.opensearch.dataprepper.plugins.otel.codec.OTelProtoCodec$OTelProtoDecoder.lambda$parseExportTraceServiceRequest$0(OTelProtoCodec.java:82) ~[otel-proto-common-2.0.1.jar:?]
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:273) ~[?:?]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682) ~[?:?]
	at org.opensearch.dataprepper.plugins.otel.codec.OTelProtoCodec$OTelProtoDecoder.parseExportTraceServiceRequest(OTelProtoCodec.java:82) ~[otel-proto-common-2.0.1.jar:?]
	at org.opensearch.dataprepper.plugins.source.oteltrace.OTelTraceGrpcService.processRequest(OTelTraceGrpcService.java:95) ~[otel-trace-source-2.0.1.jar:?]
	at org.opensearch.dataprepper.plugins.source.oteltrace.OTelTraceGrpcService.lambda$export$0(OTelTraceGrpcService.java:78) ~[otel-trace-source-2.0.1.jar:?]
	at io.micrometer.core.instrument.composite.CompositeTimer.record(CompositeTimer.java:89) ~[micrometer-core-1.9.4.jar:1.9.4]
	at org.opensearch.dataprepper.plugins.source.oteltrace.OTelTraceGrpcService.export(OTelTraceGrpcService.java:78) ~[otel-trace-source-2.0.1.jar:?]
	at io.opentelemetry.proto.collector.trace.v1.TraceServiceGrpc$MethodHandlers.invoke(TraceServiceGrpc.java:246) ~[opentelemetry-proto-1.7.1-alpha.jar:1.7.1]
	at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:182) ~[grpc-stub-1.49.0.jar:1.49.0]
	at com.linecorp.armeria.server.grpc.AbstractServerCall.invokeOnMessage(AbstractServerCall.java:374) ~[armeria-grpc-1.19.0.jar:?]
	at com.linecorp.armeria.server.grpc.AbstractServerCall.lambda$onRequestMessage$2(AbstractServerCall.java:338) ~[armeria-grpc-1.19.0.jar:?]
	at com.linecorp.armeria.internal.shaded.guava.util.concurrent.SequentialExecutor$1.run(SequentialExecutor.java:123) ~[armeria-1.19.0.jar:?]
	at com.linecorp.armeria.internal.shaded.guava.util.concurrent.SequentialExecutor$QueueWorker.workOnQueue(SequentialExecutor.java:235) ~[armeria-1.19.0.jar:?]
	at com.linecorp.armeria.internal.shaded.guava.util.concurrent.SequentialExecutor$QueueWorker.run(SequentialExecutor.java:180) ~[armeria-1.19.0.jar:?]
	at com.linecorp.armeria.common.RequestContext.lambda$makeContextAware$3(RequestContext.java:566) ~[armeria-1.19.0.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]

```
**Expected behavior**
Like in the issue description.

**Additional context**
The check of the `traceState` field happens here:
https://github.com/opensearch-project/data-prepper/blob/59c48e29a25e15b16e38f4c118477a041e458121/data-prepper-api/src/main/java/org/opensearch/dataprepper/model/trace/DefaultLink.java#L41

If you remove the links section from the trace-request.json file or add the following line to it, everything works fine:
```
"links": [
      {
          "traceId": "c0b1baf0afb0c766a31c5f96619d660e",
          "spanId": "d96ff4a5bfe1be72",
          "traceState: "nowTheFieldIsThere"
      }
```

_Originally posted by @JannikBrand in https://github.com/opensearch-project/data-prepper/issues/2089#issuecomment-1381609448_

**Comment by MrSparc** [comment]
Hi @dlvenable 
> **Additional context** The check of the `traceState` field happens here:
> 
> https://github.com/opensearch-project/data-prepper/blob/59c48e29a25e15b16e38f4c118477a041e458121/data-prepper-api/src/main/java/org/opensearch/dataprepper/model/trace/DefaultLink.java#L41
> 

TraceState is a part of Span context represented by an immutable list of string key/value pairs and formally defined by the [W3C Trace Context specification](https://www.w3.org/TR/trace-context/#tracestate-header). 



The tracestate field value is a list of list-members separated by commas (,). A list-member is a key/value pair separated by an equals sign (=). Spaces and horizontal tabs surrounding list-members are ignored. 
There can be a maximum of 32 list-members in a list.
```
3.3.1.2 list
A simple example of a list with two list-members might look like: vendorname1=opaqueValue1,vendorname2=opaqueValue2.

list  = list-member 0*31( OWS "," OWS list-member )
list-member = (key "=" value) / OWS
Identifiers for a list are short (up to 256 characters) textual identifiers.
```

From what I understand from the spec this list can be from 0 to 32 members, so 0 means it can be empty, so probably relaxing by removing the check of the `traceState` field is safe.

**Comment by MrSparc** [comment]
Pinging @dlvenable about asking  if there is an expecting roadmap version to fix this issue that makes impossible to ingest linked spans in OpenSearch because valid empty traceState values are being rejected. 
This high-priority bug was dropped from last v2.1.0 milestone.
Thanks!

**Comment by dlvenable** [comment]
@MrSparc , This should be resolved by #2363 which we will produce in a patch release 2.1.1.

**Comment by dlvenable** [comment]
This was fixed in #2363 and released in 2.1.1 and 2.2.
